### PR TITLE
Fix thread-safety issues in MetaStore and KsqlEngine

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

When running in interactive mode the `KsqlEngine` instance is shared across multiple threads, e.g. those servicing rest requests and the main statement executor thread that reads of the command topic.  However, KsqlEngine and its MetaStore are NOT currently thread-safe.  This is likely manifesting as weird errors.

This PR looks to make the MetaStore, and any state it uses, thread-safe. In addition, I've patched up the engine where I can see thread safety issues. However, this PR does not make the engine completely thread-safe as that's going to mean testing / changing a lot of code - i.e. all the classes the engine stores references to.

I've also split the FunctionRegistry into reader and writer interfaces and restricted the use of the writer interface to only those that need it.  This simplifies the code.  While in there I've made the InternalFunctionRegistry thread-safe and improved its error handling.

For what it's worth - we should favour immutable objects by default, as this gives us thread-safety for free in many situations.

### Testing done 
Added lots of tests.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java [hunk 1]
```diff
@@ -89,59 +89,61 @@ public StructuredDataSource getSource(final String sourceName) {
...
+            .getSourceForQueries()
+            .stream()
+            .collect(Collectors.joining(", "));
+
+      final String sinkForQueriesMessage = source.referentialIntegrity
+            .getSinkForQueries()
+            .stream()
+            .collect(Collectors.joining(", "));
+
+      if (!sourceForQueriesMessage.isEmpty() || !sinkForQueriesMessage.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: True dat

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java [hunk 2]
```diff
@@ -154,56 +156,55 @@ public void updateForPersistentQuery(final String queryId,
   }
 
   private void addSourceNames(final Set<String> sourceNames, final String queryId) {
-    for (final String sourceName: sourceNames) {
-      final ReferentialIntegrityTableEntry referentialIntegrity =
-          dataSourceMap.get(sourceName).referentialIntegrity;
-
-      referentialIntegrity.addSourceForQueries(queryId);
-    }
+    sourceNames.forEach(sourceName -> {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: Yep, I agree. I generally would of left a forEach like this alone. I think this was likely a artefact of refactoring, i.e. I'll of converted the forEach to a stream call because I wanted to use streams, then later removed some of the operations, so it's back to just a forEach, and then not realised.

Call this kind of thing out is fine by me.  In the new code this function doesn't exist, so its not an issue ;)

**Comment by big-andy-coates** [comment]
Thanks for the review @agavra.  I've made some more changes:

I've also split the FunctionRegistry into reader and writer interfaces and restricted the use of the writer interface to only those that need it.  This simplifies the code.  While in there I've made the InternalFunctionRegistry thread-safe and improved its error handling.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/ReadonlyMetaStore.java [hunk 3]
```diff
@@ -16,7 +16,6 @@
 
 import io.confluent.ksql.function.AggregateFunctionFactory;
 import io.confluent.ksql.function.KsqlAggregateFunction;
-import io.confluent.ksql.function.KsqlFunction;
```
Comment (rodesai) [inline ksql-metastore/src/main/java/io/confluent/ksql/metastore/ReadonlyMetaStore.java]: Not specific to the changes in this PR, but I think it would be better to implement this at the interface level by having a MetaStore interface with functions that don't support writing, and a MetaStoreWriter interface with the functions for writing. MetaStoreImpl can implement both. This way we fail at compile-time instead of run-time if we try to read when we shouldn't. Its also nicer for folks writing code to get feedback about misuse of the metastore early rather than getting an exception when testing the code out.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java [hunk 4]
```diff
@@ -89,138 +90,131 @@ public StructuredDataSource getSource(final String sourceName) {
...
+
+      streamSources(sourceNames)
+          .forEach(source -> source.referentialIntegrity.addSourceForQueries(queryId));
+      streamSources(sinkNames)
+          .forEach(source -> source.referentialIntegrity.addSinkForQueries(queryId));
     }
   }
 
   @Override
   public void removePersistentQuery(final String queryId) {
```
Comment (rodesai) [inline ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: This should take the referential integrity lock too, or you might end up with a dangling query reference in one of your sources (if you have a racing updateForPersistentQuery with the same id).

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java [hunk 5]
```diff
@@ -17,51 +17,52 @@
...
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.concurrent.ThreadSafe;
 import org.apache.kafka.connect.data.Schema;
 
+@ThreadSafe
```
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: +1

File: ksql-common/src/main/java/io/confluent/ksql/function/MutableFunctionRegistry.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+package io.confluent.ksql.function;
+
+public interface MutableFunctionRegistry extends FunctionRegistry {
+
+  /**
+   * Ensure the supplied function factory is registered
+   *
+   * @param factory the factory to register.
+   */
+  void ensureFunctionFactory(UdfFactory factory);
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-common/src/main/java/io/confluent/ksql/function/MutableFunctionRegistry.java]: I changed it from `add` as the other `add` variants throw if the thing is already present.  For this method, you can call it multiple times with the same, or equivalent, factory and it won't complete. Hence `ensure`.

I could change to `addFunctionFactoryIfAbsent`, but its a bit of a mouth full. Happy either way, but not happy with just `add`.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlContext.java [hunk 7]
```diff
@@ -46,8 +48,10 @@
   public static KsqlContext create(final KsqlConfig ksqlConfig) {
     Objects.requireNonNull(ksqlConfig, "ksqlConfig cannot be null.");
     final ServiceContext serviceContext = DefaultServiceContext.create(ksqlConfig);
+    final MutableFunctionRegistry functionRegistry = new InternalFunctionRegistry();
+    UdfLoader.newInstance(ksqlConfig, functionRegistry, ".").load();
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlContext.java]: The `KsqlContext` is our embedded mode. So it is doing the equivalent of the rest app and initializing the function register.  IMHO this is the _right_ place to be doing this.  If you still disagree, maybe you could rephrase your thinking?

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 8]
```diff
@@ -88,226 +80,257 @@ public UdfFactory getUdfFactory(final String functionName) {
...
-  public boolean addFunctionFactory(final UdfFactory factory) {
-    if (!functionNameValidator.test(factory.getName())) {
-      throw new KsqlException(factory.getName() + " is not a valid function name."
-          + " Function names must be valid java identifiers and not a KSQL reserved word"
-      );
+  public void ensureFunctionFactory(final UdfFactory factory) {
+    validateFunctionName(factory.getName());
+
+    final String functionName = factory.getName().toUpperCase();
+    if (udafs.containsKey(functionName)) {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: Not picky at all mate. This is the purpose of code reviews!

Code updated to use a CS / lock around the mutation code.

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 9]
```diff
@@ -88,226 +80,257 @@ public UdfFactory getUdfFactory(final String functionName) {
...
+
+    if (udafs.putIfAbsent(functionName, aggregateFunctionFactory) != null) {
+      throw new KsqlException("Aggregate function already registered: " + functionName);
+    }
   }
 
   @Override
   public List<UdfFactory> listFunctions() {
-    return new ArrayList<>(ksqlFunctionMap.values());
+    return new ArrayList<>(udfs.values());
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: Not code I've changed.... but I'd noticed this too.

`udfs.values()` is a Collection, not a List. So this would require changing the return value. Let me take a look...

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 10]
```diff
@@ -88,226 +80,257 @@ public UdfFactory getUdfFactory(final String functionName) {
...
+  }
+
+  private static UdfFactory builtInUdfFactory(
+      final KsqlFunction ksqlFunction,
+      final boolean internal
+  ) {
+    final UdfMetadata metadata = new UdfMetadata(
+        ksqlFunction.getFunctionName(),
+        ksqlFunction.getDescription(),
+        "Confluent",
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: If this was my code, I'd agree. This isn't.  This _should_ be done, and done for all UDFs as well. But not in this PR.

File: ksql-common/src/main/java/io/confluent/ksql/function/UdfFactory.java [hunk 11]
```diff
@@ -57,7 +57,7 @@ private void checkCompatible(final KsqlFunction ksqlFunction,
                                final List<FunctionParameter> paramTypes) {
     if (udfClass != ksqlFunction.getKudfClass()) {
       throw new KsqlException("Can't add function " + ksqlFunction
-          + " as a function with the same name exists on a different class " + udfClass);
+          + " as a function with the same name exists on a different " + udfClass);
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-common/src/main/java/io/confluent/ksql/function/UdfFactory.java]: done.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: I'm against the use of `synchronize` on method signatures. It leaks internal implementation details of the class into the API, and can cause issues if other code choosing to `synchronize(metaStore)`.  

This is why I've added `Object` lock objects in other places. However, for this instance I disagree we should add a lock around all mutable state. It's accessed independently and adding course grained concurrency control won't add anything.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java [hunk 12]
```diff
@@ -250,6 +239,18 @@ public AggregateFunctionFactory getAggregateFactory(final String functionName) {
     return functionRegistry.listAggregateFunctions();
   }
 
+  private Stream<SourceInfo> streamSources(final Set<String> sourceNames) {
+    return sourceNames.stream()
+        .map(sourceName -> {
+          final SourceInfo sourceInfo = dataSources.get(sourceName);
+          if (sourceInfo == null) {
+            throw new KsqlException("Unknown source: " + sourceName);
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: I'm not worried about the overhead of checking each source exists _twice_. (Not four times, as there are two different sets of streams and each is checked twice). 

Nor am I worried about lazy stream evaluation or throwing exceptions from streams. 

Yes, we could throw a message with all the missing streams, but I think the ROI is limited. The check never even existed before and we've not had any issues reported. Calls would previously have failed with an NPE.  This exception is thrown if a source exists when a statement is first being executed, but is dropped before the new query is registered.  

In short, I prefer this more succinct code that explicit checking up front.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/ReadonlyMetaStore.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/ReadonlyMetaStore.java]: Sure, like I've one for `FunctionRegistry` you mean?  Yep, but one for another PR.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java (hunk 4 above)
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: done.

File: ksql-common/src/main/java/io/confluent/ksql/function/UdfFactory.java [hunk 13]
```diff
@@ -57,7 +57,7 @@ private void checkCompatible(final KsqlFunction ksqlFunction,
                                final List<FunctionParameter> paramTypes) {
     if (udfClass != ksqlFunction.getKudfClass()) {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-common/src/main/java/io/confluent/ksql/function/UdfFactory.java]: Agreed, this will need addressing before we can support loading functions on demand.

File: ksql-common/src/main/java/io/confluent/ksql/function/MutableFunctionRegistry.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * Register the supplied {@code ksqlFunction}.
+   *
+   * <p>Note: a suitable function factory must already have been register via
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-common/src/main/java/io/confluent/ksql/function/MutableFunctionRegistry.java]: done.

File: ksql-common/src/main/java/io/confluent/ksql/function/MutableFunctionRegistry.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-common/src/main/java/io/confluent/ksql/function/MutableFunctionRegistry.java]: I've also updated the Java docs with more details.

**Comment by big-andy-coates** [comment]
Thanks for the reviews @agavra / @rodesai 

> * If we really don't want to fully synchronize, we could simplify the synchronization by restricting the access model. Both MetaStore and FunctionRegistry have mutable and immutable interfaces. At minimum we want full synchronization of a thread using the mutable interfaces with threads using the immutable interfaces, but not to other threads using the mutable interface (since only one thread should ever have a mutable reference to a given instance of either). This would let us drop the referential integrity lock for instance.

Yep, that's true. Though the risk is always that someone comes along who doesn't understand the access model and does the wrong think, and this is the missed in a code review.

IMHO, better to have the classes thread-safe in general and only worry about removing the overhead if it is an issue.

> * Another way to simplify the model would be to just not be thread-safe at all. KsqlResource can just do all its work against a sandbox.

That's certainly worth considering...

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlContext.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlContext.java]: Ah! I didn't realize that the class I was looking at that calls this is a test class. Makes sense to me now.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: Yeah, I'm not worried about the performance either - it was just an added benefit. I still prefer the "up front" approach (I think it may actually end up being fewer lines of code) but it's purely subjective so we can defer to your approach since you're doing the hard coding work 😝

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: I'm ambivalent about whether we use the synchronized on the method signatures or just use a synchronized block, but we shouldn't ever allow for mixing state and synchronization by locking on something like `metastore` - that's criminal! 

Interesting point, though, about it leaking into the API, didn't really think about that before. Would that still be a concern if the method is only accessed via an interface that doesn't specify `synchronized`?

[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: > It leaks internal implementation details of the class into the API

I don't buy this - we're actually losing some of the benefits of encapsulation: readability and being less bug-prone. It's also arguable that its leaking into the API. The interface defines the API, so it would seem fine for the implementation classes to use synchronized.

> can cause issues if other code choosing to `synchronize(metaStore)`

Can you elaborate on this point? synchronized is reentrant so it should be a problem to use metastore from inside `synchronize(metaStore)` somehow. Are you worried about getting deadlocked somehow?

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/ReadonlyMetaStore.java (hunk 3 above)
Comment (rodesai) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/ReadonlyMetaStore.java]: > Sure, like I've one for `FunctionRegistry` you mean?

Yeah exactly - I reviewed the metastore changes first :D

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: > Interesting point, though, about it leaking into the API, didn't really think about that before. Would that still be a concern if the method is only accessed via an interface that doesn't specify `synchronized`?

As long as the type is only ever accessed via the interface, yeah, then that's better - but people can still do things like `synchonrize(metaStore)`.

[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: > Are you worried about getting deadlocked somehow?

Yes, I've seen weird A-B dead locking issues before, hence I always sync on an private object.

[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: > I don't buy this - we're actually losing some of the benefits of encapsulation:

Not sure how this is related to encapsulation...

[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java]: > It's also debatable that its leaking into the API - the interface defines the API

In this instance, yes.  I was talking more in general terms.

My other concern stands - i,e. it's a safer pattern to use private locks.
